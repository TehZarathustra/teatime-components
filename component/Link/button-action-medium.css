@value font-medium, padding-medium from '../const.css';

.link
{
  composes: button from '../Button/button-action.css';

  padding: padding-medium;

  font: font-medium;
}

.disabled
{
  composes: disabled from '../Button/button-action.css';
}
