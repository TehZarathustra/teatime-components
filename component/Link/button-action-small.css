@value font-small, padding-small from '../const.css';

.link
{
  composes: button from '../Button/button-action.css';

  padding: padding-small;

  font: font-small;
}

.disabled
{
  composes: disabled from '../Button/button-action.css';
}
