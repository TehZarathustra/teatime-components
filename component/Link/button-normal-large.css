@value font-large, padding-large from '../const.css';

.link
{
  composes: button from '../Button/button-normal.css';

  padding: padding-large;

  font: font-large;
}

.disabled
{
  composes: disabled from '../Button/button-normal.css';
}
