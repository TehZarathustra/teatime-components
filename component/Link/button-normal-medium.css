@value font-medium, padding-medium from '../const.css';

.link
{
  composes: button from '../Button/button-normal.css';

  padding: padding-medium;

  font: font-medium;
}

.disabled
{
  composes: disabled from '../Button/button-normal.css';
}
