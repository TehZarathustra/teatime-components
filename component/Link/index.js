import {noop} from '../../lib/util';
import React, {Component} from 'react';
import classNames from 'classnames';

import buttonActionLarge from './button-action-large.css';
import buttonActionMedium from './button-action-medium.css';
import buttonActionSmall from './button-action-small.css';
import buttonNormalLarge from './button-normal-large.css';
import buttonNormalMedium from './button-normal-medium.css';
import buttonNormalSmall from './button-normal-small.css';
import linkCommonLarge from './link-common-large.css';
import linkCommonMedium from './link-common-medium.css';
import linkCommonSmall from './link-common-small.css';
import linkCornflowerBlueLarge from './link-cornflower-blue-large.css';
import linkCornflowerBlueMedium from './link-cornflower-blue-medium.css';
import linkCornflowerBlueSmall from './link-cornflower-blue-small.css';

type Props = {
  children?: any,
  className?: string,
  disabled: boolean,
  download: boolean | string,
  href: string,
  id?: string,
  onClick: Function,
  rel?: string,
  size: 'l' | 'm' | 's',
  style?: Object,
  styles?: Object,
  target: '_blank' | '_parent' | '_self' | '_top',
  theme: 'action' | 'blue' | 'common' | 'normal',
  type?: string,
};

type LinkProps = {};

const cssModules = {
  'action-l': buttonActionLarge,
  'action-m': buttonActionMedium,
  'action-s': buttonActionSmall,
  'blue-l': linkCornflowerBlueLarge,
  'blue-m': linkCornflowerBlueMedium,
  'blue-s': linkCornflowerBlueSmall,
  'common-l': linkCommonLarge,
  'common-m': linkCommonMedium,
  'common-s': linkCommonSmall,
  'normal-l': buttonNormalLarge,
  'normal-m': buttonNormalMedium,
  'normal-s': buttonNormalSmall,
};

class Link extends Component {
  props: Props

  static defaultProps: {
    disabled: boolean,
    download: boolean,
    onClick: Function,
    size: 'l' | 'm' | 's',
    target: '_blank' | '_parent' | '_self' | '_top',
    theme: 'action' | 'blue' | 'common' | 'normal',
  }

  css: {
    link: string,
  }

  constructor(props: Props) {
    super(props);

    this.css = props.styles || cssModules[props.theme + '-' + props.size];

    (this:any)._onClick = this._onClick.bind(this);
  }

  componentWillReceiveProps(nextProps: Props) {
    this.css = props.styles || cssModules[nextProps.theme + '-' + nextProps.size];
  }

  _computeLinkProps(): LinkProps {
    const {
      children,
      className,
      disabled,
      download,
      href,
      id,
      rel,
      target,
      type,
    } = this.props;
    const {css} = this;
    // https://www.jitbit.com/alexblog/256-targetblank---the-most-underestimated-vulnerability-ever/
    return {
      children,
      className: classNames(css.link, {
        [css.disabled]: disabled,
      }, className),
      disabled,
      download,
      href,
      id,
      onClick: this._onClick,
      rel: typeof rel === 'undefined' && target === '_blank'
        ? 'noopener noreferrer'
        : rel,
      target,
      type,
    };
  }

  _onClick(e: SyntheticEvent) {
    const {disabled, onClick} = this.props;

    if (!disabled) onClick(e)
    else e.preventDefault();
  }

  render() {
    const linkProps = this._computeLinkProps();

    return (
      <a {...linkProps}/>
    );
  }
}

Link.defaultProps = {
  disabled: false,
  download: false,
  onClick: noop,
  size: 'm',
  target: '_self',
  theme: 'common',
};

export default Link;
