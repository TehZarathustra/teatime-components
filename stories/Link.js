'use strict';

const { action, storiesOf } = require('@kadira/storybook');
const { delimiter } = require('./stories.css');
const Link = require('../component/Link');
const React = require('react');

storiesOf('Link', module)
  .add('theme: `action`', () => (
    <div>
      size: `xs`{' '}
      <Link
        onClick={action('click')}
        size='xs'
        theme='action'>
        Link
      </Link>

      <hr className={delimiter}/>

      size: `s`{' '}
      <Link
        onClick={action('click')}
        size='s'
        theme='action'>
        Link
      </Link>

      <hr className={delimiter}/>

      size: `m`{' '}
      <Link
        onClick={action('click')}
        size='m'
        theme='action'>
        Link
      </Link>

      <hr className={delimiter}/>

      size: `l`{' '}
      <Link
        onClick={action('click')}
        size='l'
        theme='action'>
        Link
      </Link>
    </div>
  ))
  .add('theme: `link`', () => (
    <div>
      size: `xs`{' '}
      <Link
        onClick={action('click')}
        size='xs'
        theme='link'>
        Link
      </Link>

      <hr className={delimiter}/>

      size: `s`{' '}
      <Link
        onClick={action('click')}
        size='s'
        theme='link'>
        Link
      </Link>

      <hr className={delimiter}/>

      size: `m`{' '}
      <Link
        onClick={action('click')}
        size='m'
        theme='link'>
        Link
      </Link>

      <hr className={delimiter}/>

      size: `l`{' '}
      <Link
        onClick={action('click')}
        size='l'
        theme='link'>
        Link
      </Link>
    </div>
  ))
  .add('theme: `normal`', () => (
    <div>
      size: `xs`{' '}
      <Link
        onClick={action('click')}
        size='xs'
        theme='normal'>
        Link
      </Link>

      <hr className={delimiter}/>

      size: `s`{' '}
      <Link
        onClick={action('click')}
        size='s'
        theme='normal'>
        Link
      </Link>

      <hr className={delimiter}/>

      size: `m`{' '}
      <Link
        onClick={action('click')}
        size='m'
        theme='normal'>
        Link
      </Link>

      <hr className={delimiter}/>

      size: `l`{' '}
      <Link
        onClick={action('click')}
        size='l'
        theme='normal'>
        Link
      </Link>
    </div>
  ));
