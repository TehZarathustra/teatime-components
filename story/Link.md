### Theme `action`

```
<div>
  size: `xs` <Link size='xs' theme='action'>Link</Link>

  <hr className='delimiter'/>

  size: `s` <Link size='s' theme='action'>Link</Link>

  <hr className='delimiter'/>

  size: `m` <Link size='m' theme='action'>Link</Link>

  <hr className='delimiter'/>

  size: `l` <Link size='l' theme='action'>Link</Link>
</div>
```


### Theme `link`

```
<div>
  size: `xs` <Link size='xs' theme='link'>Link</Link>

  <hr className='delimiter'/>

  size: `s` <Link size='s' theme='link'>Link</Link>

  <hr className='delimiter'/>

  size: `m` <Link size='m' theme='link'>Link</Link>

  <hr className='delimiter'/>

  size: `l` <Link size='l' theme='link'>Link</Link>
</div>
```


### Theme `normal`

```
<div>
  size: `xs` <Link size='xs' theme='normal'>Link</Link>

  <hr className='delimiter'/>

  size: `s` <Link size='s' theme='normal'>Link</Link>

  <hr className='delimiter'/>

  size: `m` <Link size='m' theme='normal'>Link</Link>

  <hr className='delimiter'/>

  size: `l` <Link size='l' theme='normal'>Link</Link>
</div>
```
