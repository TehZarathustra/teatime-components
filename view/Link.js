'use strict';

const React = require('react');
const { Component, PropTypes } = React;
const cssModules = require('react-css-modules');

class Link extends Component {
  render() {
    return (
      <a {...this.props} styleName='control'></a>
    );
  }
}

Link.defaultProps = {
  styles: {},
};

Link.propTypes = {
  href: PropTypes.string.isRequired,
  onBlur: PropTypes.func,
  onClick: PropTypes.func,
  onContextMenu: PropTypes.func,
  onDoubleClick: PropTypes.func,
  onFocus: PropTypes.func,
  onKeyDown: PropTypes.func,
  onKeyPress: PropTypes.func,
  onKeyUp: PropTypes.func,
  onMouseDown: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  onMouseMove: PropTypes.func,
  onMouseOut: PropTypes.func,
  onMouseOver: PropTypes.func,
  onMouseUp: PropTypes.func,
  onTouchCancel: PropTypes.func,
  onTouchEnd: PropTypes.func,
  onTouchMove: PropTypes.func,
  onTouchStart: PropTypes.func,
  styles: PropTypes.object,
};

module.exports = cssModules(Link);
