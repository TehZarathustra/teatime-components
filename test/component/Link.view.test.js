'use strict';

import Link from '../../component/Link';
import React from 'react';
import renderer from 'react-test-renderer';

test('defaultProp #size should have a value `m`', () => {
  expect(Link.defaultProps.size).toBe('m');
});

test('defaultProp #theme should have a value `common`', () => {
  expect(Link.defaultProps.theme).toBe('common');
});

test('Renders correctly with default props', () => {
  const tree = renderer.create(
    <Link href='#'>
      Link Content
    </Link>
  );

  expect(tree).toMatchSnapshot();
});

test('Renders correctly with defined props', () => {
  const tree = renderer.create(
    <Link
      className='outer'
      disabled
      download={true}
      href='#'
      id='links id'
      target='_blank'
      type='video/mp4'>
      Link Content
    </Link>
  ).toJSON();

  expect(tree).toMatchSnapshot();
});
