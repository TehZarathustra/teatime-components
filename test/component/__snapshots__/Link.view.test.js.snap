// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders correctly with default props 1`] = `
<a
  className="link"
  disabled={false}
  download={false}
  href="#"
  id={undefined}
  onClick={[Function]}
  rel={undefined}
  target="_self"
  type={undefined}
>
  Link Content
</a>
`;

exports[`Renders correctly with defined props 1`] = `
<a
  className="link disabled outer"
  disabled={true}
  download={true}
  href="#"
  id="links id"
  onClick={[Function]}
  rel="noopener noreferrer"
  target="_blank"
  type="video/mp4"
>
  Link Content
</a>
`;
